Loading the libraries and dataset

In [2]:
import pandas as pd
import matplotlib.pyplot as plt

In [28]:
df = pd.read_csv("C:\\Users\\USER\\Downloads\\insurance.csv")

In [29]:
df

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

Finding the missing values

In [30]:
df.isna().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

In [31]:
df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

It shows parameters about the dataset
1) Age : Around 25% of the dataset around the age of 27 years, 50% of the dataset assimilates the information about people who are less than 39 years old and 75% of the dataset covers people aged less than 51 years 
2) bmi: So, 25% of the dataset has their bmi below 26 units, 50% or the median value of the dataset have clocked the value 30 units for the emausremnt of bmi and 75% of the dataset have bmi value less 34 units. 
3) children: So 25% of the dataset have no children, 50 % of the dataset has atleast 1 child while 75% of the dataset have 2 children.
4) charges: 25% of people were charged less than 4740 units of curency, 50% were charged less than 9382 units of currency, and 75% were charged less than 16639 units of currency. 

Since we have categorical variables we do some encoding and replace the values for better analysis and modelling puprose

In [32]:
df['sex'].unique()

array(['female', 'male'], dtype=object)

In [33]:
df['sex']=df['sex'].map({'female':0,'male':1})

In [34]:
df.head()

age  sex     bmi  children smoker     region      charges
0   19    0  27.900         0    yes  southwest  16884.92400
1   18    1  33.770         1     no  southeast   1725.55230
2   28    1  33.000         3     no  southeast   4449.46200
3   33    1  22.705         0     no  northwest  21984.47061
4   32    1  28.880         0     no  northwest   3866.85520

In [35]:
df['smoker'].unique()

array(['yes', 'no'], dtype=object)

In [36]:
df['smoker']=df['smoker'].map({'yes':1,'no':0})

In [37]:
df.head()

age  sex     bmi  children  smoker     region      charges
0   19    0  27.900         0       1  southwest  16884.92400
1   18    1  33.770         1       0  southeast   1725.55230
2   28    1  33.000         3       0  southeast   4449.46200
3   33    1  22.705         0       0  northwest  21984.47061
4   32    1  28.880         0       0  northwest   3866.85520

In [38]:
df['region'].unique()

array(['southwest', 'southeast', 'northwest', 'northeast'], dtype=object)

In [39]:
df['region']=df['region'].map({'southwest':1,'southeast':2,'northwest':3,'northeast':4})

In [40]:
df

age  sex     bmi  children  smoker  region      charges
0      19    0  27.900         0       1       1  16884.92400
1      18    1  33.770         1       0       2   1725.55230
2      28    1  33.000         3       0       2   4449.46200
3      33    1  22.705         0       0       3  21984.47061
4      32    1  28.880         0       0       3   3866.85520
...   ...  ...     ...       ...     ...     ...          ...
1333   50    1  30.970         3       0       3  10600.54830
1334   18    0  31.920         0       0       4   2205.98080
1335   18    0  36.850         0       0       2   1629.83350
1336   21    0  25.800         0       0       1   2007.94500
1337   61    0  29.070         0       1       3  29141.36030

[1338 rows x 7 columns]

Keeping Feature Matrix in X and target variable "Charge" in the y variable

In [42]:
df.columns

Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges'], dtype='object')

In [43]:
X = df.drop(['charges'],axis=1)

In [44]:
X

age  sex     bmi  children  smoker  region
0      19    0  27.900         0       1       1
1      18    1  33.770         1       0       2
2      28    1  33.000         3       0       2
3      33    1  22.705         0       0       3
4      32    1  28.880         0       0       3
...   ...  ...     ...       ...     ...     ...
1333   50    1  30.970         3       0       3
1334   18    0  31.920         0       0       4
1335   18    0  36.850         0       0       2
1336   21    0  25.800         0       0       1
1337   61    0  29.070         0       1       3

[1338 rows x 6 columns]

In [45]:
y = df['charges']

In [46]:
y

0       16884.92400
1        1725.55230
2        4449.46200
3       21984.47061
4        3866.85520
           ...     
1333    10600.54830
1334     2205.98080
1335     1629.83350
1336     2007.94500
1337    29141.36030
Name: charges, Length: 1338, dtype: float64

Splitting the dataset into training and test dataset

In [47]:
from sklearn.model_selection import train_test_split

In [48]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2, random_state=42)

In [49]:
y_test

764      9095.06825
887      5272.17580
890     29330.98315
1293     9301.89355
259     33750.29180
           ...     
109     47055.53210
575     12222.89830
535      6067.12675
543     63770.42801
846      9872.70100
Name: charges, Length: 268, dtype: float64

Since we have more than one independent variables we use different models to test which one gives better fitted value and predicts most accurately on the baiss of the past dataset available

Importing the Models

In [50]:
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVR
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor

In [51]:
lr = LinearRegression()
lr.fit(X_train,y_train)
svm = SVR()
svm.fit(X_train,y_train)
rf = RandomForestRegressor()
rf.fit(X_train,y_train)
gr = GradientBoostingRegressor()
gr.fit(X_train,y_train)

GradientBoostingRegressor()

In [52]:
y_pred1 = lr.predict(X_test)
y_pred2 = svm.predict(X_test)
y_pred3 = rf.predict(X_test)
y_pred4 = gr.predict(X_test)

In [53]:
df = pd.DataFrame({'Actual':y_test,'Lr':y_pred1,'svm':y_pred2,'rf':y_pred3,'gr':y_pred4})

In [54]:
df

Actual            Lr          svm            rf            gr
764    9095.06825   8924.407244  9548.261584  11385.711440  11001.128629
887    5272.17580   7116.295018  9492.515425   5184.614460   5840.174656
890   29330.98315  36909.013521  9648.758701  28229.060629  28001.980112
1293   9301.89355   9507.874691  9555.044136  11191.917474   9745.291602
259   33750.29180  27013.350008  9420.421978  34595.033144  33639.100981
...           ...           ...          ...           ...           ...
109   47055.53210  39116.968669  9648.902852  47431.962851  45431.423211
575   12222.89830  11814.555568  9625.431547  12550.486184  12465.025294
535    6067.12675   7638.107736  9504.168517   6399.538195   6974.336525
543   63770.42801  40959.081722  9605.004594  46958.794487  47862.047791
846    9872.70100  12258.228529  9590.987268  10120.913409  10289.655388

[268 rows x 5 columns]

In [55]:
from sklearn import metrics

In [56]:
score1 = metrics.r2_score(y_test, y_pred1)
score2= metrics.r2_score(y_test,y_pred2)
score3= metrics.r2_score(y_test,y_pred3)
score4= metrics.r2_score(y_test,y_pred4)

In [57]:
print(score1, score2, score3, score4)

0.7833463107364539 -0.07229762787861826 0.8622024678629909 0.8779726251291786


Here gradient Boosting Regressor works much better than the models in predicting the nearby values for the target dependent variable 'charge' so we use it for further preidction

Predict charges for new customer

In [58]:
data = {'age':40,
         'sex':1,
           'bmi':40.30,
            'children':4,
             'smoker':1,
              'region':2}
df = pd.DataFrame(data,index=[0])
df

age  sex   bmi  children  smoker  region
0   40    1  40.3         4       1       2

In [59]:
new_pred = gr.predict(df)
print(new_pred)

[43931.13613049]


So the fee charged from a new customer on the baiss of certain available information is around 43931 units of currecy value